Tribler/Core/TorrentChecker/session.py

```
@@ -198,309 +185,384 @@ def __init__(self, tracker_url, tracker_address, announce_page, on_result_callba
+                #  - https://wiki.theory.org/BitTorrentSpecification#Tracker_.27scrape.27_Convention
+                seeders = complete
                 leechers = incomplete
 
-                # handle the retrieved information
-                self._on_result_callback(infohash, seeders, leechers)
+                # Store the information in the dictionary
+                seed_leech_dict[infohash] = (seeders, leechers)
 
                 # remove this infohash in the infohash list of this session
                 if infohash in unprocessed_infohash_list:
                     unprocessed_infohash_list.remove(infohash)
 
         elif 'failure reason' in response_dict:
-            self._logger.debug(u"%s Failure as reported by tracker [%s]", self, repr(response_dict['failure reason']))
-
-            return False
+            self._logger.info(u"%s Failure as reported by tracker [%s]", self, repr(response_dict['failure reason']))
+            self.failed()
+            return
 
         # handle the infohashes with no result (seeders/leechers = 0/0)
         for infohash in unprocessed_infohash_list:
             seeders, leechers = 0, 0
-            # handle the retrieved information
-            self._on_result_callback(infohash, seeders, leechers)
-        return True
+            seed_leech_dict[infohash] = (seeders, leechers)
 
+        self._is_finished = True
+        self.result_deferred.callback(seed_leech_dict)
 
-class UdpTrackerSession(TrackerSession):
+    def cleanup(self):
+        """
+        Cleans the session by cancelling all deferreds and closing sockets.
+        :return: A deferred that fires once the cleanup is done.
+        """
+        super(HttpTrackerSession, self).cleanup()
+        if self.request is not None:
+            self.request.cancel()
+
+        if self.result_deferred:
+            self.result_deferred.cancel()
```

The code you linked is equivalent of a catch all try/except, you are swallowing any error that may happen on the callback chain.

Are you sure that's the only thing you need to do when canceling that callback chain?

I think the whole flow could be streamlined a bit. I'll comment on the code.